text/3064-bitfield.md

```diff
@@ -0,0 +1,384 @@
+
+A field named `_` is called an *unnamed* field. All other fields are called
+*named* fields. Each named field annotated with `bitfield(N)` occupies `N` bits
+of storage. Unnamed fields do not occupy any storage.
+
+When reading and writing a bit-field with type `bool`, `bool` is treated like
+`u1` with `true` corresponding to `1` and `false` corresponding to `0`.
+
+When a field annotated with `bitfield(N)` is read, the value has the type
+of the field and the behavior is as follows:
+
+- The `N` bits of storage occupied by the bit-field are read.
+- If the type of the field is signed (resp. unsigned), the bits are 1-extended
+  (resp. 0-extended) to the size of the type of the field. (1-extended means
+  that the new bits will have the value of the most significant bit. In
+  particular, bit-fields with signed types with width 1 can only store the
+  values `0` and `-1`.)
+- The resulting bits are interpreted as a value of the type of the field.
+
+When a field annotated with `bitfield(N)` is written, the value to be written
+must have the type of the field and the behavior is as follows:
+
+- If overflow checks are enabled and the value is outside the range of values
+  that can be read from the field, the overflow check fails.
+- The bitmask `(1 << N) - 1` is applied to the value and the remaining `N`
+  significant bits are written to the storage of the bit-field.
+
+If the overflow check is performed at compile time, the behavior is subject to
+the `arithmetic_overflow` lint.
+
+By the *layout* of a struct containing bit-fields, we mean the following
+properties:
+
+- The size and alignment of the struct.
+- The offsets of all non-bit-field fields.
+- For each named bit-field, the bits of the object representation used as its
+  storage.
+
+The language reference shall document for each target the layout of structs
+containing bit-fields.
+
+The intended behavior is that the layout is the same layout as produced by Clang
+when compiling the corresponding C struct for the same target. The corresponding
+C struct is constructed as follows:
+
+- Translate the struct to the corresponding C struct as if the `bitfield`
+  annotations were not present and as if `_` were a valid identifier for regular
+  fields.
+- For each field that has a `bitfield(N)` annotation in the Rust struct, append
+  `: N` to the declaration in the C struct.
+- For each field in the C struct whose name is `_`, delete the field name.
+
+The `&` and `&mut` operators cannot be applied to bit-fields.
+
+This implies that, in order to (mutably) access a bit-field, one must have a
+(mutable) reference to the struct. Therefore, no concurrent accesses of the
+struct are possible unless all accesses are immutable. Therefore, unlike in C,
+the compiler can implement access to bit-fields with any kind of load/write
+instruction, even if such an instruction overlaps the memory locations of other
+fields in the struct.
```

Comment:
I'd prefer not to leave the behavior quite this unspecified. I would prefer to have it explicitly specified that a store to a bitfield is permitted to be combined with other stores to bitfields in a single operation, but that the operation will not be performed via a wider write than a single machine word *unless* the entire structure is actually being written.

People do use bitfields for flags within larger structures, where other parts of the structure might be concurrently accessed while under a lock, and they assume that a write to one of those bitfields will not overwrite a lock-protected value. We should be precise *enough* that people know how they can avoid that.

Reply:
>People do use bitfields for flags within larger structures, where other parts of the structure might be concurrently accessed while under a lock

You're right, I didn't take into account that other fields might have interior mutability. I'll replace this by the usual language:

The memory locations of two different fields within a struct do not overlap if

- At least one of them is not a bit-field
- They are separated, in declaration order, by at least one field
	- that is not a bit-field or
	- that is a bit-field with width 0.

"Memory location" is the language used by C to determine when two unsynchronized concurrent accesses are well-defined.

Reply:
That sounds reasonable to me.

Reply:
I've chosen a somewhat different formulation to avoid pulling in the C memory model:

When a bit-field field is accessed, the abstract machine may also access
adjacent bit-field fields but not fields that are separated from the field by a
StructBodyElement that is not a bit-field field. (Note: This paragraph restricts
the kinds of loads and stores the compiler can perform when accessing a
bit-field. This paragraph does not need to be specially advertised to users as
the inability to take references to bit-field fields makes it impossible to
access adjacent bit-field fields in otherwise sound code.)